lib/spack/spack/util/path.py

```diff
@@ -320,40 +321,59 @@ def substitute_path_variables(path):
     return path
 
 
-def _get_padding_string(length):
-    spack_path_padding_size = len(SPACK_PATH_PADDING_CHARS)
-    num_reps = int(length / (spack_path_padding_size + 1))
-    extra_chars = length % (spack_path_padding_size + 1)
-    reps_list = [SPACK_PATH_PADDING_CHARS for i in range(num_reps)]
-    reps_list.append(SPACK_PATH_PADDING_CHARS[:extra_chars])
-    return os.path.sep.join(reps_list)
-
-
-def add_padding(path, length):
-    """Add padding subdirectories to path until total is length characters
-
-    Returns the padded path. If path is length - 1 or more characters long,
-    returns path. If path is length - 1 characters, warns that it is not
-    padding to length
-
-    Assumes path does not have a trailing path separator"""
-    padding_length = length - len(path)
-    if padding_length == 1:
-        # The only 1 character addition we can make to a path is `/`
-        # Spack internally runs normpath, so `foo/` will be reduced to `foo`
-        # Even if we removed this behavior from Spack, the user could normalize
-        # the path, removing the additional `/`.
-        # Because we can't expect one character of padding to show up in the
-        # resulting binaries, we warn the user and do not pad by a single char
-        tty.warn("Cannot pad path by exactly one character.")
-    if padding_length <= 0:
+def add_padding(path: str, length: int, fsencode: Callable[[str], bytes] = os.fsencode):
+    """Given a path ``path``, add placeholder subdirectories such that its length in code
+    units (depenending on the fs encoding) *without* trailing directory seperators is at
+    least ``length``.
+
+    This should work with UTF-16 encoding of paths, and also deal properly with paths containing
+    code points that are multiple code units. In principle since PEP 529 we could assume code
+    units are one byte, but unfortunately UTF-8 paths are not strictly enforced on Windows.
+
+    Returns: padded path as string, without trailing separator.
+    """
+
+    # Drop trailing dir seperators
+    path = path.rstrip(os.path.sep)
+
+    # NOTE: the assumption is that "." is one unit
+    bytes_per_unit = len(fsencode("."))
+
+    # How many units is our current path, and how many should we add.
+    units_in_path = len(fsencode(path)) // bytes_per_unit
+    pad_units = length - units_in_path
```

Comment:
Wouldn't `len(path)` be better for getting the current units, since `utf-8` can be variable-width?

```suggestion
    pad_units = length - len(path)
```

Reply:
In general `len(path) <= units_in_path`; in utf-8 a code point can be 1 to 4 units, and every unit is one byte. So with the suggestion the effective length in bytes would be longer than prescribed in `padded_length`, which can be problematic when `padded_length` is large and the store dir contains many multibyte chars

Reply:
Ah got it, I wasn't picking up on `len` reporting code points instead of code units.

Reply:
That's the part that makes it a pain ;p